### PR TITLE
Refactor: read metrics from config.ros for pmlog_summary

### PR DESCRIPTION
**Pull request by xiangce**
- Update pmlog_summary to read all "mandatory on" metrics in config.ros
- Move the depended specs to the new datasource directory
- Add test for these datasources

Signed-off-by: Xiangce Liu <xiangceliu@redhat.com>

### All Pull Requests:

Check all that apply:

* [x] Have you followed the guidelines in our Contributing document, including the instructions about commit messages?
* [ ] Is this PR to correct an issue?
* [x] Is this PR an enhancement?

### Complete Description of Additions/Changes:

- Update pmlog_summary to read all "mandatory on" metrics in config.ros
- Move the depended specs to the new datasource directory
- Add test for these datasources

### CODE REVIEW

**Comment by xiangce** [comment]
For the ROS rules, currently, some of the metrics, e.g. the `mem.util.available`, are still not collected by the `pmlog_summary` spec. 

Since the RosConfig already specified what should be mandatorily enabled in the pmloger, do you think that we directly collect the `pmlogsummary` as these mandatory metrics by default? 

By doing this, when there are new metrics that should be collected by the `pmlog_summary` spec, we can only easily add them to the `config.ros` file. How do you think?

@gnurag, @apuntamb, @kgaikwad - please have a look.

Thanks.

File: insights/specs/datasources/pcp.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    Returns:
+        str: Full path to the latest pmlogger file
+
+    Raises:
+        SkipComponent: When no metrics get from the RosConfig
+    """
+    ros = broker[RosConfig]
+    metrics = set()
+
+    for spec in ros.specs:
```
Comment (kgaikwad) [inline insights/specs/datasources/pcp.py:76]: One query - do we need to handle `null` condition here when ROS config file absent on targeted system?

**Review by kgaikwad — COMMENTED** [review]
> Since the RosConfig already specified what should be mandatorily enabled in the pmloger, do you think that we directly collect the pmlogsummary as these mandatory metrics by default?
> By doing this, when there are new metrics that should be collected by the `pmlog_summary` spec, we can only easily add them to the `config.ros` file. How do you think?

@xiangce, good idea! :+1: 
Yes, with current implementation there is no difference between metrics listed under `config.ros` & metrics passed in `pmlogsummary` command so make sense to me. Haven't tested but code wise looks good to me except one inline query.

**Comment by apuntamb** [comment]
@xiangce Yeah makes sense to me too since many new metrics should now be available in both pmlogsummary and config.ros parsers. For this sync, it's better to be able to easily add those metrics collected by pmlogsummary to the config.ros.

File: insights/specs/datasources/pcp.py (hunk 1 above)
Comment (xiangce) [reply insights/specs/datasources/pcp.py]: @kgaikwad, Thanks for reviewing.

Since the `RosConfig` is a `Required` component of the datasource, when the `Required` component is not triggered, this datasoruce will cannot be trigger too.

See: https://insights-core.readthedocs.io/en/latest/api_index.html#insights.core.dr.ComponentType

TLDR;, see the following test:

```python
# sudo mv /var/lib/pcp/config/pmlogger/config.ros /var/lib/pcp/config/pmlogger/config.ros.bak
# ls /var/lib/pcp/config/pmlogger/config.ros
ls: cannot access '/var/lib/pcp/config/pmlogger/config.ros': No such file or directory
# insights-inspect insights.specs.Specs.pmlog_summary
Missing Dependencies:
    At Least One Of:
        insights.specs.default.DefaultSpecs.pmlog_summary
# ^^ The spec will not get fired when missing the `config.ros`
# sudo mv /var/lib/pcp/config/pmlogger/config.ros.bak /var/lib/pcp/config/pmlogger/config.ros
# ^^ Get the `config.ros` file back
# insights-inspect insights.specs.Specs.pmlog_summary

IPython Console Usage Info:

Enter 'pmlog_summary.' and tab to get a list of properties
Example:
In [1]: pmlog_summary.<property_name>
Out[1]: <property value>

To exit ipython enter 'exit' and hit enter or use 'CTL D'

Starting IPython Interpreter Now

In [1]: !ls /var/lib/pcp/config/pmlogger/config.ros
/var/lib/pcp/config/pmlogger/config.ros

In [2]: pmlog_summary.content
Out[2]:
['disk.all.total  0.401 count / sec',
 'kernel.all.cpu.idle  3.973 none',
 'kernel.all.cpu.nice  0.003 none',
 'kernel.all.cpu.steal  0.000 none',
 'kernel.all.cpu.sys  0.005 none',
 'kernel.all.cpu.user  0.012 none',
 'kernel.all.cpu.wait.total  0.000 none',
 'mem.physmem  7835748.000 Kbyte',
 'mem.util.bufmem  5236.000 Kbyte',
 'mem.util.cached  2572533.667 Kbyte',
 'mem.util.free  690659.853 Kbyte',
 'mem.util.used  7145088.147 Kbyte']
```

Comment (kgaikwad) [reply insights/specs/datasources/pcp.py]: Thank you @xiangce for the details.

**Comment by apuntamb** [comment]
@bfahr Hello! With respect to these changes, do we want to change the parser logic in RosConfig or pmlog_summary parsers? For e.g. the following metrics may be introduced now:
```json
 "metrics": {
    "disk.dev.total": {
      "vda": 3.377
    },
    "hinv.ncpu": 1,
    "kernel.all.cpu.idle": 0.976,
    "mem.physmem": 2018447.992,
    "mem.util.used": 1592813.593,
    "kernel.all.pressure.cpu.some.avg": {
      "10 second": 0.74,
      "1 minute": 0.526,
      "5 minute": 0.467
    },
    "kernel.all.pressure.io.full.avg": {
      "10 second": 0.015,
      "1 minute": 0.009,
      "5 minute": 0.001
    },
    "kernel.all.pressure.io.some.avg": {
      "10 second": 0.05,
      "1 minute": 0.032,
      "5 minute": 0.018
    },
    "kernel.all.pressure.memory.full.avg": {
      "10 second": 0,
      "1 minute": 0,
      "5 minute": 0
    },
    "kernel.all.pressure.memory.some.avg": {
      "10 second": 0,
      "1 minute": 0,
      "5 minute": 0
    }
  }
```
Since the metrics look a bit different/complex could you guide if we will need to change our parser logic?

**Comment by bfahr** [comment]
> @bfahr Hello! With respect to these changes, do we want to change the parser logic in RosConfig or pmlog_summary parsers? For e.g. the following metrics may be introduced now:
> 
> ```json
>  "metrics": {
>     "disk.dev.total": {
>       "vda": 3.377
>     },
>     "hinv.ncpu": 1,
>     "kernel.all.cpu.idle": 0.976,
>     "mem.physmem": 2018447.992,
>     "mem.util.used": 1592813.593,
>     "kernel.all.pressure.cpu.some.avg": {
>       "10 second": 0.74,
>       "1 minute": 0.526,
>       "5 minute": 0.467
>     },
>     "kernel.all.pressure.io.full.avg": {
>       "10 second": 0.015,
>       "1 minute": 0.009,
>       "5 minute": 0.001
>     },
>     "kernel.all.pressure.io.some.avg": {
>       "10 second": 0.05,
>       "1 minute": 0.032,
>       "5 minute": 0.018
>     },
>     "kernel.all.pressure.memory.full.avg": {
>       "10 second": 0,
>       "1 minute": 0,
>       "5 minute": 0
>     },
>     "kernel.all.pressure.memory.some.avg": {
>       "10 second": 0,
>       "1 minute": 0,
>       "5 minute": 0
>     }
>   }
> ```
> 
> Since the metrics look a bit different/complex could you guide if we will need to change our parser logic?

@apuntamb if the input data is changing then the parser will need to be modified to parser the new data.  I would start by adding tests with any new data to see if the parser still works.

**Review by bfahr — CHANGES_REQUESTED** [review]

File: insights/specs/datasources/pcp.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        pm_file = "/var/log/pcp/pmlogger/ros/%s.index" % (pcp_log_date)
+        try:
+            if not (os.path.exists(pm_file) and os.path.isfile(pm_file)):
+                raise SkipComponent("No such file: {0}".format(pm_file))
+        except Exception as e:
+            raise SkipComponent("Failed to check for pmlogger file existence: {0}".format(str(e)))
+
+        metrics = set()
+        ros = broker[RosConfig]
+        for spec in ros.specs:
```
[the remark this answers is not in the document]
Comment (xiangce) [reply insights/specs/datasources/pcp.py:65]: @bfahr - Thanks for the heads up. Updated.

File: insights/specs/datasources/pcp.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            raise SkipComponent("No pmlogger file: {0}".format(pm_file))
+
+    except Exception as e:
+        raise SkipComponent("Failed to check pmlogger file existence: {0}".format(str(e)))
+
+    metrics = set()
+    try:
+        ros = broker[RosConfig]
+        for spec in ros.specs:
+            if spec.get('state') == 'mandatory on':
```
Comment (xiangce) [inline insights/specs/datasources/pcp.py:66]: @kgaikwad, @apuntamb - I just checked the latest `config.ros` and did find any `mandatory on` metrics are specified. Please help confirm if it's okay to apply this condition to this spec. Since with the latest `config.ros` nothing will be collected by the pmlogsummary. Thanks.

**Comment by xiangce** [comment]
> > @bfahr Hello! With respect to these changes, do we want to change the parser logic in RosConfig or pmlog_summary parsers? For e.g. the following metrics may be introduced now:
> > ```json
> >  "metrics": {
> >     "disk.dev.total": {
> >       "vda": 3.377
> >     },
> >     "hinv.ncpu": 1,
> >     "kernel.all.cpu.idle": 0.976,
> >     "mem.physmem": 2018447.992,
> >     "mem.util.used": 1592813.593,
> >     "kernel.all.pressure.cpu.some.avg": {
> >       "10 second": 0.74,
> >       "1 minute": 0.526,
> >       "5 minute": 0.467
> >     },
> >     "kernel.all.pressure.io.full.avg": {
> >       "10 second": 0.015,
> >       "1 minute": 0.009,
> >       "5 minute": 0.001
> >     },
> >     "kernel.all.pressure.io.some.avg": {
> >       "10 second": 0.05,
> >       "1 minute": 0.032,
> >       "5 minute": 0.018
> >     },
> >     "kernel.all.pressure.memory.full.avg": {
> >       "10 second": 0,
> >       "1 minute": 0,
> >       "5 minute": 0
> >     },
> >     "kernel.all.pressure.memory.some.avg": {
> >       "10 second": 0,
> >       "1 minute": 0,
> >       "5 minute": 0
> >     }
> >   }
> > ```
> > Since the metrics look a bit different/complex could you guide if we will need to change our parser logic?
> 
> @apuntamb if the input data is changing then the parser will need to be modified to parse the new data. I would start by adding tests with any new data to see if the parser still works.

@apuntamb - I think it's necessary to check if the output of the `pmlogsummary` will change or not after adding the new metrics to the `config.ros`. Would you please help check with?

File: insights/specs/datasources/pcp.py (hunk 1 above)
Comment (xiangce) [reply insights/specs/datasources/pcp.py]: That's what I missed for datasource, and according to Bob's comment, I used `try/except` for this check. Sorry for the incomplete explanation.

File: insights/specs/datasources/pcp.py (hunk 3 above)
Comment (kgaikwad) [reply insights/specs/datasources/pcp.py:66]: > I just checked the latest config.ros on my test system and did not find any mandatory on metrics are specified. 

@xiangce - did you mean new metrics here? 
For reference - ROS side PR available to add new metrics under `mandatory on` -  https://github.com/RedHatInsights/ros-backend/pull/123. 
In latest config file change also, ROS do have old metrics under `mandatory on`  - https://github.com/RedHatInsights/ros-backend/blob/main/ansible-playbooks/ros_install_and_set_up.yml#L44
so I want to confirm that.

**Review by kgaikwad — CHANGES_REQUESTED** [review]
Just few nitpicks, feel free to ignore them :-)

File: insights/specs/datasources/pcp.py (hunk 3 above)
Comment (xiangce) [reply insights/specs/datasources/pcp.py:66]: Great, Thanks for the confirmation! @kgaikwad

File: insights/specs/datasources/pcp.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    Determines the pmlogger file and the metrics to collect via `pmlog_summary`
+
+    Returns:
+        str: Full arguments string that will be passed to the `pmlog_summary`,
+             which contains the `pmloger` file and the required `metrics`.
+
+    Raises:
+        SkipComponent: Raises when meeting one of the following scenario:
+                       - No pmlogger process is running
+                       - No pmloger file
```
[the remark this answers is not in the document]
Comment (xiangce) [reply insights/specs/datasources/pcp.py]: Fixed

File: insights/specs/datasources/pcp.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    try:
+        ps = broker[Ps]
+        if not ps.search(COMMAND_NAME__contains='pmlogger'):
+            raise SkipComponent("No 'pmlogger' is running")
+
+        pcp_log_date = (datetime.date.today() - datetime.timedelta(days=1)).strftime("%Y%m%d")
+        pm_file = "/var/log/pcp/pmlogger/ros/{0}.index".format(pcp_log_date)
+
+        if not (os.path.exists(pm_file) and os.path.isfile(pm_file)):
+            raise SkipComponent("No pmlogger file: {0}".format(pm_file))
```
[the remark this answers is not in the document]
Comment (xiangce) [reply insights/specs/datasources/pcp.py]: Fixed

File: insights/specs/datasources/pcp.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+
+@datasource(Ps, RosConfig, HostContext)
+def pmlog_summary_args(broker):
+    """
+    Determines the pmlogger file and the metrics to collect via `pmlog_summary`
+
+    Returns:
+        str: Full arguments string that will be passed to the `pmlog_summary`,
+             which contains the `pmloger` file and the required `metrics`.
```
[the remark this answers is not in the document]
Comment (xiangce) [reply insights/specs/datasources/pcp.py]: Updated. Thanks!

**Comment by apuntamb** [comment]
> @apuntamb - I think it's necessary to check if the output of the pmlogsummary will change or not after adding the new metrics to the config.ros. Would you please help check with?

@xiangce Yes surely. The ros-backend side PR is in review. Also, to check the output, as suggested by @bfahr I will open a PR that updates the tests of RosConfig parser, how does that sound to you?

**Comment by xiangce** [comment]
> @xiangce Yes surely. The ros-backend side PR [RedHatInsights/ros-backend#123](https://github.com/RedHatInsights/ros-backend/pull/123) is in review. Also, to check the output, as suggested by @bfahr I will open a PR that updates the tests of RosConfig parser, how does that sound to you?
@apuntamb, I'm totally Okay with that. Thanks.

**Comment by xiangce** [comment]
FYI: PR#3290 enhanced the `PmLogSummary` for the new metrics.

**Comment by apuntamb** [comment]
@xiangce Tested locally along with #3290 and works as expected. 
Currently, I think RosConfig parser is not breaking when I tested locally based on updated metrics in config.ros. But, if there is any chance it can fail, please if you could throw some light on the same, that'll be great.
